libopae++/src/handle.cpp

```diff
@@ -33,41 +34,102 @@ namespace fpga
 namespace types
 {
 
-handle::handle(fpga_handle h)
+handle::handle(fpga_handle h, uint32_t mmio_region, uint8_t *mmio_base)
 : handle_(h)
+, mmio_region_(mmio_region)
+, mmio_base_(mmio_base)
 {
 }
 
 handle::~handle(){
     close();
 }
 
+bool handle::write(uint64_t offset, uint32_t value)
+{
+    return FPGA_OK == fpgaWriteMMIO32(handle_, mmio_region_, offset, value);
+}
+
+bool handle::write(uint64_t offset, uint64_t value)
+{
+    return FPGA_OK == fpgaWriteMMIO64(handle_, mmio_region_, offset, value);
+}
+
+bool handle::read(uint64_t offset, uint32_t &value) const
+{
+    return FPGA_OK == fpgaReadMMIO32(handle_, mmio_region_, offset, &value);
+}
+
+bool handle::read(uint64_t offset, uint64_t &value) const
+{
+    return FPGA_OK == fpgaReadMMIO64(handle_, mmio_region_, offset, &value);
+}
+
+handle::ptr_t handle::open(fpga_token token, int flags, uint32_t mmio_region){
+    fpga_handle c_handle = nullptr;
+    uint8_t *mmio_base = nullptr;
+    ptr_t p;
 
-handle::ptr_t handle::open(fpga_token token, int flags){
-    fpga_handle c_handle;
     auto res = fpgaOpen(token, &c_handle, flags);
-    if (res == FPGA_OK){
-        return handle::ptr_t(new handle(c_handle));
+    if (res == FPGA_OK) {
+
+        // This will fail for FPGA_DEVICE (no mmio region), so we need to continue when it does.
+        res = fpgaMapMMIO(c_handle, mmio_region, reinterpret_cast<uint64_t **>(&mmio_base));
+
+        if (res != FPGA_OK) {
```

Comment:
Do we want to use the c++ API to get properties and check the type?
Also, wouldn't we want to call 'fpgaMapMMIO` only if type is `FPGA_ACCELERATOR`?